File: observalog-node/src/wire.ts
import { Err, Outcome } from './fields';
import { KEY_TO_ABBREV } from './dict';
import { PART_A_SCHEMA_VERSION, PARENT_SPAN_ABSENT } from './wire-contract';

export interface LogEntry {
    // Part A
    traceId:      string;
    spanId:       string;
    parentSpan:   string;
    seq:          number;
    serviceCode:  number;
    levelCode:    number;
    outcomeCode:  number;
    tsMs:         number;
    // Part B
    event:        string;
    message:      string;
    durationMs:   number | undefined;
    outcome:      Outcome | undefined;
    error:        Err | undefined;
    userId:       string;
    journeyStage: string;
    ctx:          Record<string, unknown>;
}

// EncodePartA produces the 55-byte fixed-position header line.
// Matches Go: fmt.Sprintf("A:%d|%-14s|%-7s|%-7s|%02x|%d|%d|%d|%013d\n", ...)
export function encodePartA(entry: LogEntry): string {
    const parentSpan = entry.parentSpan || PARENT_SPAN_ABSENT;
    return (
        `A:${PART_A_SCHEMA_VERSION}|` +
        `${entry.traceId.substring(0, 14).padEnd(14)}|` +
        `${entry.spanId.substring(0, 7).padEnd(7)}|` +
        `${parentSpan.substring(0, 7).padEnd(7)}|` +
        `${(entry.seq & 0xff).toString(16).padStart(2, '0')}|` +
        `${entry.serviceCode}|` +
        `${entry.levelCode}|` +
        `${entry.outcomeCode}|` +
        `${entry.tsMs.toString().padStart(13, '0')}\n`
    );
}

// EncodePartB produces the abbreviated JSON payload line.
export function encodePartB(entry: LogEntry): string {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const out: Record<string, any> = {};
    out['e'] = entry.event;
    out['m'] = entry.message;
    if (entry.durationMs !== undefined)  out['ms'] = entry.durationMs;
    if (entry.outcome !== undefined)     out['o']  = entry.outcome;

    if (entry.error) {
        out['er'] = {
            ek: entry.error.kind,
            ec: entry.error.code,
            em: entry.error.message,
            rt: entry.error.retryable,
        };
    }

    const ctxKeys = Object.keys(entry.ctx);
    if (ctxKeys.length > 0) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const ctx: Record<string, any> = {};
        for (const [k, v] of Object.entries(entry.ctx)) {
            ctx[KEY_TO_ABBREV[k] ?? k] = v;
        }
        out['c'] = ctx;
    }

    if (entry.userId)       out['ui'] = entry.userId;
    if (entry.journeyStage) out['js'] = entry.journeyStage;

    return JSON.stringify(out) + '\n';
}

File: observalog-java/src/main/java/com/observalog/Wire.java
package com.observalog;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.Map;

public final class Wire {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private Wire() {}

    // EncodePartA produces the 55-byte fixed-position header line.
    // Matches Go: fmt.Sprintf("A:%d|%-14s|%-7s|%-7s|%02x|%d|%d|%d|%013d\n", ...)
    public static byte[] encodePartA(LogEntry entry) {
        String parentSpan = entry.parentSpan.isEmpty() ? WireContract.PARENT_SPAN_ABSENT : entry.parentSpan;
        String line = String.format(Locale.US,
            "A:%d|%-14s|%-7s|%-7s|%02x|%d|%d|%d|%013d\n",
            WireContract.PART_A_SCHEMA_VERSION,
            truncate(entry.traceId, 14),
            truncate(entry.spanId, 7),
            truncate(parentSpan, 7),
            entry.seq & 0xFF,
            entry.serviceCode,
            entry.levelCode,
            entry.outcomeCode,
            entry.tsMs
        );
        return line.getBytes(java.nio.charset.StandardCharsets.UTF_8);
    }

    // EncodePartB produces the abbreviated JSON payload line.
    public static byte[] encodePartB(LogEntry entry) throws IOException {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("e", entry.event);
        out.put("m", entry.message);
        if (entry.durationMs != null)          out.put("ms", entry.durationMs);
        if (entry.outcome != null)             out.put("o", entry.outcome.wire);

        if (entry.error != null) {
            Map<String, Object> err = new LinkedHashMap<>();
            err.put("ek", entry.error.kind);
            err.put("ec", entry.error.code);
            err.put("em", entry.error.message);
            err.put("rt", entry.error.retryable);
            out.put("er", err);
        }

        if (!entry.ctx.isEmpty()) {
            Map<String, Object> ctx = new LinkedHashMap<>();
            for (Map.Entry<String, Object> e : entry.ctx.entrySet()) {
                String abbrev = Dict.KEY_TO_ABBREV.get(e.getKey());
                ctx.put(abbrev != null ? abbrev : e.getKey(), e.getValue());
            }
            out.put("c", ctx);
        }

        if (!entry.userId.isEmpty())       out.put("ui", entry.userId);
        if (!entry.journeyStage.isEmpty()) out.put("js", entry.journeyStage);

        byte[] json = MAPPER.writeValueAsBytes(out);
        byte[] line = new byte[json.length + 1];
        System.arraycopy(json, 0, line, 0, json.length);
        line[json.length] = '\n';
        return line;
    }

    private static String truncate(String s, int n) {
        return s.length() > n ? s.substring(0, n) : s;
    }
}
